### PR TITLE
Conversions between Value, PathBuf and OsString

### PR DESCRIPTION
**Pull request by nagisa**
I did an initial take on this. I can see a number of ways to improve on this PR, but all of the things seem fairly involved.

### CODE REVIEW

File: glib/src/translate.rs [hunk 1]
```diff
@@ -604,6 +604,14 @@ impl<'a> ToGlibPtr<'a, *const c_char> for Path {
         let tmp = path_to_c(self);
         Stash(tmp.as_ptr(), tmp)
     }
+
+    #[inline]
+    fn to_glib_full(&self) -> *const c_char {
+        let stash = self.to_glib_none();
+        unsafe {
+            ffi::g_strdup(stash.0) as *mut c_char
```
Comment (nagisa) [inline glib/src/translate.rs:612]: Since `to_glib_none` allocates anyway, it seems like it'd make more sense to implement `to_glib_none` in terms of `to_glib_full`, and not the other way around, potentially saving an allocation.

Comment (sdroege) [reply glib/src/translate.rs:612]: Good catch, that's true (also for all the other string-y types). The main problem is that we then have to replicate the `NUL`-check of `CString`, or silently truncate at `NUL`. For now continuing with the current behaviour should be safest (and I think is also the correct behaviour here).

Do you want to add another commit on top of this to implement this?

File: glib/src/translate.rs [hunk 2]
```diff
@@ -682,6 +731,11 @@ impl<'a> ToGlibPtr<'a, *mut c_char> for OsString {
         let tmp = os_str_to_c(self);
         Stash(tmp.as_ptr() as *mut c_char, tmp)
     }
+
+    #[inline]
+    fn to_glib_full(&self) -> *mut c_char {
+        OsStr::to_glib_full(&*self)
+    }
 }
```
Comment (sdroege) [inline glib/src/translate.rs:739]: Ideally all the changes that add more FFI conversions for these types should be a separate commit

File: glib/src/types.rs [hunk 3]
```diff
@@ -351,6 +351,10 @@ builtin!(f32, F32);
 builtin!(f64, F64);
 builtin!(str, STRING);
 builtin!(String, STRING);
+builtin!(std::ffi::OsString, STRING);
+builtin!(std::ffi::OsStr, STRING);
+builtin!(std::path::PathBuf, STRING);
+builtin!(std::path::Path, STRING);
```
Comment (sdroege) [inline glib/src/types.rs:357]: It's really sad that GLib has annotations for distinguishing `utf8` from `filename` strings but at the type system level (`GType`) it's all a string.

Fortunately it doesn't matter too much because a) who uses non-UTF8 locales on UNIXy systems nowadays and b) filenames are always UTF-8 in GLib on Windows.

File: glib/src/translate.rs (hunk 1 above)
Comment (sdroege) [reply glib/src/translate.rs:612]: Well there's also the Windows special case here: in GLib they're all UTF-8 but in Rust [they're](https://doc.rust-lang.org/std/ffi/struct.OsString.html)

> On Windows, strings are often arbitrary sequences of non-zero 16-bit values, interpreted as UTF-16 when it is valid to do so.

So that conversion also needs to be moved around a bit.

**Comment by sdroege** [comment]
Looks generally good to me, thanks :) You'll also want to run `cargo fmt` once though.

> I can see a number of ways to improve on this PR, but all of the things seem fairly involved.

What other ways than the above do you see? Some of the duplication could be reduced by macros but that doesn't seem worth it for these few lines.

**Review by GuillaumeGomez — APPROVED** [review]
Seems good to me. I'm curious about the improvements you have in mind though!

**Comment by nagisa** [comment]
So one thing is that I'd love for this conversion to also transparently work when the `Value` contains a `GString` type. Another expressed as the inline comment – I think it'd be great to reduce the amount of allocations we end up with during the conversion to the C string `Value`. Finally, panics (instead of `Result`s) when the string e.g. contains WTF-8 that can't easily become UTF-8 bother me a little, but changing that would be fairly hard AFAICT.

**Comment by nagisa** [comment]
Ran cargo fmt and split up the changes into more self-contained commits.

**Comment by sdroege** [comment]
> Writing the commit message for `impl SetValue/FromValue for OsStr{,ing}` I began wondering if we really want to impl `OsStr{,ing}` variants right now. In particular I suspect that its going to block us from making these represented as `GString` by default in the future.
> > These correspond to UTF-8 encoded C string Value type currently, as an
> > arbitrary choice. A better option would be a GString, which can contain
> > arbitrary byte strings, thus potentially supporting UTF-16 encoded data
> > of Windows as well.

In GLib on Windows, all these strings (paths, environment variables, ...) are actually UTF-8 so that's correct. On UNIXy systems it should already pass those through in whatever encoding they're in, both here and in GLib.

What problems do you see?

**Comment by sdroege** [comment]
> What problems do you see?

@nagisa How should we proceed here? :)

**Comment by nagisa** [comment]
I don't know, honestly.

From this PR it is very apparent to me that the string situation is a big mess and it is way over my head to resolve. I would love it if interior null bytes within a `OsString` did not explode in some way, but it is also pretty apparent to me that people would end up being surprised if a `GString` came out from that kind of conversion.

I'll leave a decision up to you all on how to proceed here. Feel free to build on my PR, implement something else entirely or whatever else – you all have been working on glib/gtk/etc for far longer than I have been and likely know better than me what's the best course of action is here.

**Comment by sdroege** [comment]
I think the right thing to do here is to implement the `Value` traits for `PathBuf` and `OsString` here like you did.

The overall string situation is really hard to use and generally a problem in all C/FFI situations with Rust and I would consider that out of scope for this PR at least. I'm not sure how to solve that nicely either, other than panicking on interior `\0` and similar runtime solutions.

**Comment by GuillaumeGomez** [comment]
The main issue here is that Rust has specific types whereas glib doesn't. Even though it's not great to have `GString` for everything, I think it's the best way for now. The next thing we can do (since we have conversions thanks to this PR) would be to enforce that each function/method is returning the appropriate Rust type (like `PathBuf` for pathes, etc).

**Comment by sdroege** [comment]
> The main issue here is that Rust has specific types whereas glib doesn't. Even though it's not great to have `GString` for everything, I think it's the best way for now. The next thing we can do (since we have conversions thanks to this PR) would be to enforce that each function/method is returning the appropriate Rust type (like `PathBuf` for pathes, etc).

You can't use a `GString` for a `glib::Value` that contains a path/filename though. That's why we would need the value trait impls that are part of this PR.

File: glib/src/translate.rs [hunk 4]
```diff
@@ -652,6 +674,12 @@ impl<'a> ToGlibPtr<'a, *const c_char> for OsStr {
         let tmp = os_str_to_c(self);
         Stash(tmp.as_ptr(), tmp)
     }
+
+    #[inline]
+    fn to_glib_full(&self) -> *const c_char {
+        let stash = self.to_glib_none();
+        unsafe { ffi::g_strdup(stash.0) as *mut c_char }
```
Comment (GuillaumeGomez) [inline glib/src/translate.rs:681]:
```suggestion
        unsafe { ffi::g_strdup(stash.0) as *const c_char }
```

File: glib/src/value.rs [hunk 5]
```diff
@@ -1162,6 +1222,51 @@ mod tests {
...
+
+        let path = Path::new("apple");
+        let pathbuf = path.to_owned();
+        let path_value = path.to_value();
+        let pathbuf_value = pathbuf.to_value();
+
+        assert_eq!(pathbuf_value.get::<PathBuf>(), Ok(Some(pathbuf.clone())));
+        assert_eq!(path_value.get::<PathBuf>(), Ok(Some(pathbuf.clone())));
+        // Drop the backing storage, and compare with str to ensure valid memory management.
+        drop(pathbuf);
```
Comment (GuillaumeGomez) [inline glib/src/value.rs:1237]: Even though it's dropped, the memory can remain accessible (depending on the page allocation, etc). Not sure if this part of the test is that useful... Or maybe you could simply update the path value?

File: glib/src/value.rs [hunk 6]
```diff
@@ -1162,6 +1222,51 @@ mod tests {
...
+        use std::ffi::{OsStr, OsString};
+        let osstr = OsStr::new("banana");
+        let osstring = osstr.to_owned();
+        let osstr_value = osstr.to_value();
+        let osstring_value = osstring.to_value();
+
+        assert_eq!(osstring_value.get::<OsString>(), Ok(Some(osstring.clone())));
+        assert_eq!(osstr_value.get::<OsString>(), Ok(Some(osstring.clone())));
+        // Drop the backing storage, and compare with str to ensure valid memory management.
+        drop(osstring);
```
Comment (GuillaumeGomez) [inline glib/src/value.rs:1259]: Same as above.

File: glib/src/value.rs (hunk 5 above)
Comment (sdroege) [reply glib/src/value.rs:1237]: `valgrind` would complain, and at least I'm regularly running our tests with `valgrind`

Comment (GuillaumeGomez) [reply glib/src/value.rs:1237]: `valgrind` doesn't work on mac and it's really annoying to make it run with `cargo test`. Updating the value and checking that it was updated seems simpler.

Comment (sdroege) [reply glib/src/value.rs:1237]: Yes

**Comment by GuillaumeGomez** [comment]
Needs to be re-opened on gtk-rs-core.
